riscv/src/code_gen.rs

```diff
@@ -319,43 +375,154 @@ fn preamble<T: FieldElement>(runtime: &Runtime, with_bootloader: bool) -> String
+    reg val1;
+    reg val2;
+    col witness val1_col;
+    col witness val2_col;
+    col witness val3_col;
+    col witness val4_col;
+
+    col witness XX, XX_inv, XXIsZero;
+    std::utils::force_bool(XXIsZero);
+    XXIsZero * XX = 0;
+
+    // HACK: This constraint cannot be active globally, because when
+    // XX is not constrained, witgen will try to set XX, XX_inv and XXIsZero
+    // to zero, which fails this constraint. Therefore, we have to activate
+    // constrained whenever XXIsZero is used.
+    // XXIsZero = 1 - XX * XX_inv
+
     // ============== control-flow instructions ==============
 
-    instr load_label l: label -> X { X = l }
+    instr load_label X, l: label
+        link ~> regs.mstore(X, STEP, val1_col)
+    {
+        val1_col = l
+    }
 
-    instr jump l: label -> Y { pc' = l, Y = pc + 1}
-    instr jump_dyn X -> Y { pc' = X, Y = pc + 1}
+    instr jump l: label, W
+        link ~> regs.mstore(W, STEP, val3_col)
+    {
+        pc' = l,
+        val3_col = pc + 1
+    }
+    
+    instr jump_dyn X, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> regs.mstore(W, STEP, val3_col)
+    {
+        pc' = val1_col,
+        val3_col = pc + 1
+    }
 
-    instr branch_if_nonzero X, l: label { pc' = (1 - XIsZero) * l + XIsZero * (pc + 1) }
-    instr branch_if_zero X, l: label { pc' = XIsZero * l + (1 - XIsZero) * (pc + 1) }
+    instr branch_if_nonzero X, Y, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - val2_col,
+        pc' = (1 - XXIsZero) * l + XXIsZero * (pc + 1)
+    }
+
+    instr branch_if_zero X, Y, Z, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - (val2_col + Z),
+        pc' = XXIsZero * l + (1 - XXIsZero) * (pc + 1)
+    }
 
     // Skips Y instructions if X is zero
-    instr skip_if_zero X, Y { pc' = pc + 1 + (XIsZero * Y) }
+    //instr skip_if_zero X, Y { pc' = pc + 1 + (XIsZero * Y) }
+    instr skip_if_zero X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - val2_col + Z,
+        pc' = pc + 1 + (XXIsZero * W)
+    }
 
     // input X is required to be the difference of two 32-bit unsigend values.
     // i.e. -2**32 < X < 2**32
-    instr branch_if_positive X, l: label {
-        X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
+    instr branch_if_positive X, Y, Z, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        (val1_col - val2_col + Z) + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
```

Comment:
Maybe this is too late, but why not use e.g. `a` for `val1_col`? I think the name is a bit unwieldy

Reply:
agree we could rename it to something better, but I'd prefer something more unique than just "a".
How about `reg1` etc?